### PR TITLE
[Reviewer: Ellie] Chronos API changes - allowing multiples of tags

### PR DESCRIPTION
**Pull request by johadalin**
I've added the unit tests that cover all the changes i've added, but i'm unsure if there are areas that i've missed, or things that should be covered more thoroughly. 

All unit tests are again passing, and i'm back to the levels of coverage there previously were. I've run live tests against a deployment using t his code, and they also appear to pass. I will do some more detailed manual testing, sending in timers and checking the logs, to make sure that the code is behaving as expected.

### CODE REVIEW

File: src/test/test_timer.cpp [hunk 1]
```diff
@@ -254,6 +254,78 @@ TEST_F(TestTimer, FromJSONTests)
   cwtest_reset_time();
 }
 
+TEST_F(TestTimer, FromJSONTestsBadStatisticsObject)
```
Comment (eleanor-merry) [inline src/test/test_timer.cpp]: For each test can you add a comment explaining what the test is for?

File: src/main/timer.cpp [hunk 2]
```diff
@@ -679,26 +688,61 @@ Timer* Timer::from_json_obj(TimerID id,
...
+
+            // If above checks pass, Increment map entry for the tags by count.
+            if (add_tags)
+            {
+              timer->tags[(type.GetString())] += count;
+            }
+          }
+          else
+          {
+            TRC_DEBUG("Tag type not present or badly formed, or count present but badly formed. Discarding some tags.");
```
Comment (eleanor-merry) [inline src/main/timer.cpp]: Line length

File: src/main/timer.cpp [hunk 3]
```diff
@@ -679,26 +688,61 @@ Timer* Timer::from_json_obj(TimerID id,
...
-          // All tags get store within the timer as a vector
-          JSON_ASSERT_STRING(*it);
-          timer->tags.push_back(std::string(it->GetString(), it->GetStringLength()));
+          // All correctly formed tags get stored in the timer as a map of tag to count.
+          bool add_tags = true;
+
+          // Default tag count if no value is found in the JSON object.
+          uint32_t count = 1;
+
+          // Check that the tag is a string. If not, set add_tags to false.
```
Comment (eleanor-merry) [inline src/main/timer.cpp]: This feels more complicated than it needs to be. Also, you're setting a value in the if statement - I'm not sure this is what you want.

File: src/main/timer_handler.cpp [hunk 4]
```diff
@@ -663,46 +663,77 @@ void TimerHandler::save_tombstone_information(Timer* t, Timer* existing)
...
+      // Check if the new tag is not present in old_tags
+      if (old_tags.count(it->first) == 0)
+      {
+        // Not present in old_tags. Add tags to tags_to_add
+        tags_to_add[it->first] = it->second;
+      }
+      else
+      {
+        // If new_tag count is greater, add difference to tags_to_add
+        if (it->second > old_tags[it->first])
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: You can save the old_tags value in a temp variable rather than getting it each time.

**Comment by eleanor-merry** [comment]
Generally looks good, a few comments inline

**Comment by johadalin** [comment]
Markups made. 
Can you just double check the new logic in timer.cpp:691 ?

File: src/main/timer.cpp [hunk 5]
```diff
@@ -679,26 +688,69 @@ Timer* Timer::from_json_obj(TimerID id,
...
+        // Iterate over tag-info array, pulling out all valid tags and counts.
+        for (rapidjson::Value::ValueIterator it = tag_info.Begin();
+                                                  it != tag_info.End();
                                                   ++it)
         {
-          // All tags get store within the timer as a vector
-          JSON_ASSERT_STRING(*it);
-          timer->tags.push_back(std::string(it->GetString(), it->GetStringLength()));
+          // Default tag count if no value is found in the JSON object.
+          uint32_t count = 1;
```
Comment (eleanor-merry) [inline src/main/timer.cpp]: It makes more sense to define this when you're checking the count value

File: src/main/timer.cpp [hunk 6]
```diff
@@ -679,26 +688,69 @@ Timer* Timer::from_json_obj(TimerID id,
...
+               (it->HasMember("type")))
+          {
+            // Check that the "type" is a string.
+            // If not, discard this tag, and move to next tag-info object.
+            if (!((*it)["type"]).IsString())
+            {
+              TRC_DEBUG("Tag type badly formed. Discarding some tags.");
+              continue;
+            }
+            rapidjson::Value& type = (*it)["type"];
```
Comment (eleanor-merry) [inline src/main/timer.cpp]: Do you need this?

**Pull request merged**